feat: Add CG, TN, and line search algorithm that will be used in OFDFT.

sunliang98 opened the pull request:
Add CG method, Truncated-Newton, and line search algorithm and corresponding unittests to source/module_esolver/opt/, these methods will be used in OFDFT later.

dyzheng committed:
Fix: 1. only init wfc once for MD and relax; 2. reset PW_DIAG_THR every Ionic step. 3. 17 test cases reference updated

dyzheng committed:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into develop

sunliang98 committed:
<feat>: Add CG method, Truncated-Newton, and line search algorithm and corresponding unittest to source/module_esolver/opt/, these methods will be used in OFDFT later.

dyzheng committed:
Merge pull request #864 from dyzheng/develop

Fix: 1. only init wfc once for MD and relax; 2. reset PW_DIAG_THR every Ionic step. 3. 17 test cases reference updated

dyzheng committed:
Fix: 2D-block allolation error

dyzheng committed:
Fix : don't print cube file if out_chg is 0

dyzheng committed:
Merge pull request #868 from dyzheng/develop

Fix: 1, 2D-block allocation error. 2, not print charge density cube file if out_chg is 0

ddhhss committed:
Fix：Added support for makefile to module_surchem. (#871)

caic99 commented on source/module_esolver/opt/opt_DCsrch.cpp:
```diff
@@ -0,0 +1,750 @@
+extern "C"{
+/* Copyright (C) 1996 Jorge J. More'.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
```
This source file is using a different license from ABACUS.

sunliang98 replied:
I have removed this file temporarily. Later it will be replaced.

caic99 commented on source/module_esolver/opt/test/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.20)
+
+project(CG_test)
+project(TN_test)
```
Why do we need a independent CMake project here?

caic99 replied:
For tests, please refer to our [docs](https://github.com/deepmodeling/abacus-develop/blob/develop/docs/CONTRIBUTING.md#adding-a-unit-test).

sunliang98 replied:
I see, CG and TN have not been used in ABACUS for now, so I add this CMake project to compile the files in /source/module_esolver/opt separately. I will remove it and modify the tests to standard unittest when OFDFT module is completed.

pxlxingliang committed:
test(hsolver):add the parallel test of CG and DAVIDSON diagonalization

sunliang98 committed:
Delete opt/opt_DCsrch.cpp.

Delete opt/opt_DCsrch.cpp, whose license is different from ABACUS. Later it will be replaced.

dyzheng committed:
Merge pull request #865 from pxlxingliang/develop

test(hsolver):add the parallel test of CG and DAVIDSON diagonalization

sunliang98 committed:
Merge branch 'develop' of https://github.com/deepmodeling/abacus-develop into OFDFT

mohanchen reviewed (approved):
agree
